### DISCUSSION TITLE
[General] Failed eth_sendTransaction on forked mainnet

### DISCUSSION THREAD

**Discussion by teamspice** [zkSync-Community-Hub/zksync-developers#390] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I am forking mainnet using zksync era-test-node 0.1.0-alpha.17, then in a separate script calling eth_sendTransaction on the forked network with a simple approval transaction on 0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4 (USDC). It fails, even though read transactions such as balanceOf on the same contract work as expected.

### What is the specific issue or error you're encountering?

It fails with error "Invalid transaction data: invalid data".

### Can you share the error messages or logs you're receiving, if any?

Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32602,\"message\":\"Invalid transaction data: Invalid data\"},\"id\":46}\n", error={"code":-32602}, requestBody="{\"method\":\"eth_sendTransaction\",\"params\":[{\"to\":\"0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4\",\"value\":\"0x0\",\"data\":\"0x095ea7b30000000000000000000000006e2b76966cbd9cf4cc2fa0d76d24d5241e0abc2f0000000000000000000000000000000000000000000000000000000000989680\",\"from\":\"0xE999bb14881e48934A489cC9B35A4f9449EE87fb\",\"gas\":\"0x3635c9adc5dea00000\"}],\"id\":46,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="<>", code=SERVER_ERROR, version=web/5.7.1)

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

No

### Have you tried to isolate the problem, and if so, what were the results?

See above

### What steps have you already taken to try to resolve the issue?

None

**Comment by jillalbritton** [comment]
The error "Invalid transaction data: invalid data" typically indicates a problem with the transaction data being sent. Ensure that the transaction data is properly formatted and matches the requirements of the contract you're interacting with. Additionally, verify that the contract address and other parameters are correct.

**Reply by teamspice** [reply]
The transaction data is given above. From what I can see, it's properly formatted and matches the requirements of the contract I'm interacting with. The other parameters you are referring to are also listed above and from what I can see, are also correct.

**Comment by syringjerry** [comment]
You can interact with smart contracts on zkSync using the Python SDK by deploying contracts, calling their functions, and querying their states. You may refer to the zkSync Python SDK documentation and explore examples provided there for guidance.

**Reply by teamspice** [reply]
Why do I need to deploy a contract if I’m trying to interact with an already deployed contract on a forked network?

**Comment by dutterbutter** [comment]
@teamspice thanks for posting. Can you provide the simple script calling eth_sendTransaction that was used so I may reproduce exactly. I am also going to copy this over to the era-test-node repo as an issue.

**Reply by teamspice** [reply]
```
import { ethers } from "ethers";

// Specify the RPC URL of your Hardhat fork
const rpcUrl = ""; // e.g., "http://localhost:8545"
// Specify the chain ID of your Hardhat fork
const chainId = 31337; // Hardhat default chainId is 31337, adjust as needed

// Create a provider connected to the specified JSON-RPC URL and chain ID
const provider = new ethers.providers.JsonRpcProvider(rpcUrl, chainId);

// Define the transaction object
const transaction = {
  to: "0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4",
  value: "0x0",
  data: "0xa9059cbb000000000000000000000000981f198286e40f9979274e0876636e9144b8fb8e0000000000000000000000000000000000000000000000000000000000989680",
  from: "0x36615Cf349d7F6344891B1e7CA7C72883F5dc049",
};

// Send the transaction
async function sendTransaction() {
  try {
    const hash = await provider.send("eth_sendTransaction", [transaction]);
    console.log("Transaction hash:", hash);

    // Wait for the transaction to be mined and get the receipt
    const receipt = await provider.waitForTransaction(hash);
    console.log("Transaction receipt:", receipt);
  } catch (error) {
    console.error("Error sending transaction:", error);
  }
}

// Execute the function
sendTransaction();

```

**Reply by teamspice** [reply]
Note that this is a transfer transaction on USDC (the above issue was for approval, but this seems simpler)

**Reply by dutterbutter** [reply]
@teamspice can you see if this is still an issue with the latest era-test-node release `era-test-node@0.1.0-alpha.19`?

**Reply by teamspice** [reply]
Still an issue. With the exact same code as above, except using 0xE999bb14881e48934A489cC9B35A4f9449EE87fb as the from address, and adding a call to hardhat_impersonateAccount with 0xE999bb14881e48934A489cC9B35A4f9449EE87fb as the input parameter, the error I am receiving now is:

`Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"Execution error: Transaction HALT: virtual machine entered unexpected state. Please contact developers and provide transaction details that caused this error. Error description: Assertion error: Overhead higher than gasLimit\",\"data\":\"0x3fafbf289f1134f1a0d3e6efc05d2e33475d4453edc435a587b3525ecb15977f\"},\"id\":51}\n", error={"code":3,"data":"0x3fafbf289f1134f1a0d3e6efc05d2e33475d4453edc435a587b3525ecb15977f"}, requestBody="{\"method\":\"eth_sendTransaction\",\"params\":[{\"to\":\"0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4\",\"value\":\"0x0\",\"data\":\"0xa9059cbb00000000000000000000000042310121982db7fa65552ae797016ab63b4302920000000000000000000000000000000000000000000000000000000000989680\",\"from\":\"0xE999bb14881e48934A489cC9B35A4f9449EE87fb\",\"maxFeePerGas\":\"0x5c631f80\",\"maxPriorityFeePerGas\":\"0x59682f00\",\"gasLimit\":\"100000000000000\"}],\"id\":51,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="", code=SERVER_ERROR, version=web/5.7.1)`

**Reply by dutterbutter** [reply]
Thanks for checking - will be investigating!

**Reply by teamspice** [reply]
Any ETA on this? Should I open an issue in the era test node repo?

**Reply by dutterbutter** [reply, marked as answer]
Hey @teamspice 

My apologies on the delay here. We have a draft PR just need to resolve a few issues. Opening an issue to era-test-node would certainly be helpful so we can ensure the fixes address the bug directly. Again my apologies on the delay here.

**Reply by teamspice** [reply]
Thanks. issue created https://github.com/matter-labs/era-test-node/issues/278

**Comment by knowenlophoroger** [comment]
The user is forking the mainnet using zksync era-test-node 0.1.0-alpha.17 and calling eth_sendTransaction on a simple approval transaction on the USDC contract, but it fails with the error "Invalid transaction data: invalid data". The user has not made any recent changes to the contract, and there are no external libraries or contracts involved. The relevant portions of the contract code are not provided, and the user has not attempted to isolate the problem or resolve the issue yet. No repository link or additional details are provided.
